sbat.c

```diff
@@ -120,4 +121,168 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
+{
+	struct sbat_var *new_entry = AllocatePool(sizeof(*new_entry));
+
+	INIT_LIST_HEAD(&new_entry->list); 
+	new_entry->component_generation = comp_gen;
+	new_entry->component_name_size = comp_name_size;
+	new_entry->component_name = comp_name;
+	return new_entry;
+}
+
+EFI_STATUS
+add_entry(list_t *list, const CHAR8 *comp_gen, const CHAR8 *comp_name_size,
+          const CHAR8 *comp_name)
+{
+	struct sbat_var *new;
+	new = new_entry(comp_gen, comp_name_size, comp_name);
+	if (!new)
+		return EFI_OUT_OF_RESOURCES;
+
+	list_add_tail(&new->list, list);
+	return EFI_SUCCESS;
+}
+
+static void
+clean_up_vars(list_t *entries)
+{
+	list_t *pos = NULL, *tmp = NULL;
+	for_each_sbat_var_safe(pos, tmp, entries) {
+		struct sbat_var *entry = list_entry(pos, struct sbat_var, list);
+		list_del(&entry->list);
+		if (entry->component_generation)
+			FreePool((CHAR8 *)entry->component_generation);
+		if (entry->component_name_size)
+			FreePool((CHAR8 *)entry->component_name_size);
+		if (entry->component_name)
+			FreePool((CHAR8 *)entry->component_name);
+		FreePool(entry);
+	}
+}
+
+EFI_STATUS
+parse_sbat_var(list_t *entries)
+{
+	UINT8 *data = 0;
+	UINTN datasize, i;
+	EFI_STATUS efi_status;
+
+	INIT_LIST_HEAD(entries); 
+	efi_status = get_variable(L"SBAT", &data, &datasize, SHIM_LOCK_GUID);
+	if (EFI_ERROR(efi_status)) {
+		console_error(L"Failed to read SBAT variable",
+			      efi_status);
+		return efi_status;
+	}
+
+	CHAR8 *start = (CHAR8 *)data;
+	CHAR8 *end = (CHAR8 *)data + datasize;
+	if (is_utf8_bom(start, datasize))
+		start += 3;
+	while ((*end == '\r' || *end == '\n') && end < start)
+		end--;
+	*end = '\0';
+
+	dprint(L"SBAT variable data:\n");
+
+	char delim;
+	while (start) {
+		const CHAR8 *fields[3] = { NULL, };
+		for (i = 0; i < 4; i++) {
+			const CHAR8 *tmp;
+			/* on fourth iteration we check if we had extra stuff on line while parsing
+			   component_name. If delimeter on 3rd iteration was ',', this means that
+			   we have comments after component_name. get_sbat_field in this if condition
+			   parses comments, if they are present and drops them.
+			*/
+			if (i == 3 && start[0] != '\0') {
+				if (delim == ',')
+					start = get_sbat_field(start, end, &tmp, '\n');
+				break;
+			}
+			delim = ',';
+			/* we do not want to jump to next line and grab stuff from that
+			*/
+			if ( (strchrnula(start,'\n') - start + 1) <= (strchrnula(start,',') - start + 1)) {
+				delim = '\n';
+			}
+			if (!start) {
+				goto error;
+			}
+			start = get_sbat_field(start, end, &tmp, delim);
+			/*   to be replaced when we have strdupa()
+			*/
+			fields[i] = strndupa(tmp, strlen(tmp));
+			if (!fields[i]) {
+				goto error;
+			}
+		}
+		dprint(L"component %a with generation %a\n", fields[2], fields[0]);
+		efi_status = add_entry(entries, fields[0], fields[1], fields[2]);
+		if (EFI_ERROR(efi_status))
+			goto error;
+	}
+	FreePool(data);
+	return EFI_SUCCESS;
+error:
+	dprint(L"failed to parse SBAT variable\n");
```

Comment:
Since this is an error condition probably we don't want a debug message and instead a `perror()`.